pysmiles/smiles_helper.py

```diff
@@ -579,3 +578,78 @@ def increment_bond_orders(molecule, max_bond_order=3):
         molecule.edges[idx, jdx]['order'] = new_order
         missing_bonds[idx] -= edge_missing
         missing_bonds[jdx] -= edge_missing
+
+def mark_chiral_atoms(molecule):
+    """
+    For all nodes tagged as chiral, figure out the three
+    substiuents and annotate the node with a tuple that
+    has the order in which to rotate. This essentially
+    corresponds to the definition of an improper dihedral
+    angle centered on the chiral atom.
+    """
+    chiral_nodes = nx.get_node_attributes(molecule, 'stereo')
+    for node, direction in chiral_nodes.items():
+        # discard the node with the lowest index because we
+        # are looking down that node; it is not part of the
+        # stereo projection
+        neighbors = sorted(molecule.neighbors(node))[1:]
+        if len(neighbors) != 3:
+            n = len(neighbors) + 1
+            msg = (f"Chiral node {node} has {n} neighbors, which "
+                    "is different than the four expected for "
+                    "tetrahedral chirality. If the chiral center "
+                    "contains hydrogen you need to set explicit_hydrogen "
+                    "to True.")
```

Comment:
Why can't there be an implicit hydrogen? That would by definition get the lowest priority, IIRC

Reply:
that's not how the open smiles definition works though; the chirality is not according to R/S but a local reference frame that is arbitrary viewed from the first atom which may or may not have lowest priority

Reply:
> If one of the neighbor atoms is a hydrogen and is represented as an atomic property of the chiral center (rather than explicitly as [H]), then it is considered to be the first atom in the clockwise or anticlockwise accounting.

So if the smiles is: `N[C@H](O)Br`, then `N=0, H=1, O=2, Br=3`. i.e., look along N-C bond, and the rotation is H-O-Br

Reply:
No it is actually even more crazy:

```
N[C@H](O)Br --> here it is H0 N1 O2 Br3
```

```
N[C][H](O)Br --> here it is N0 H1 O2 Br3
```

the text reads:

*If one of the neighbor atoms is a hydrogen and is represented as an atomic property of the chiral center (rather than explicitly as [H]), then it is considered to be the first atom in the clockwise or anticlockwise accounting.* Now the question is how do we figure that out in the workflow?

Reply:
Yes, you're right.

Maybe have the central atom with implicit H itself as first "neighbour"?

Reply:
I think it is best to tag the node when it has the implicit hydrogen def. This does not change the fact that you cannot have implicit hydrogen here. The entire annotation only makes sense for when you have all hydrogen present.

Reply:
I still don't understand the "you cannot have implicit hydrogen". To me "implicit" means that the hydrogen is not an explicit node in the graph, but rather a node attribute

Reply:
I guess we should hide this chirality assignment behind a flag, which then requires explicit hydrogen atoms. Basically, the point being that the assignment of node indices makes only sense if these nodes are actually present.